libraries/standard/mqtt/integration-test/mqtt_system_stest.c

```diff
@@ -0,0 +1,718 @@
+
+    pSubscriptionList[ 0 ].qos = qos;
+    pSubscriptionList[ 0 ].pTopicFilter = pTopic;
+    pSubscriptionList[ 0 ].topicFilterLength = strlen( pTopic );
+
+    /* Generate packet identifier for the UNSUBSCRIBE packet. */
+    globalUnsubscribePacketIdentifier = MQTT_GetPacketId( pContext );
+
+    /* Send UNSUBSCRIBE packet. */
+    return MQTT_Unsubscribe( pContext,
+                             pSubscriptionList,
+                             sizeof( pSubscriptionList ) / sizeof( MQTTSubscribeInfo_t ),
+                             globalUnsubscribePacketIdentifier );
+}
+
+
+static MQTTStatus_t publishToTopic( MQTTContext_t * pContext,
+                                    const char * pTopic,
+                                    MQTTQoS_t qos )
+{
+    assert( pContext != NULL );
+    MQTTPublishInfo_t publishInfo;
+
+    /* Set the retain flag to false to avoid side-effects across test runs. */
+    publishInfo.retain = false;
+
+    publishInfo.qos = qos;
+    publishInfo.dup = false;
+    publishInfo.pTopicName = pTopic;
+    publishInfo.topicNameLength = strlen( pTopic );
+    publishInfo.pPayload = MQTT_EXAMPLE_MESSAGE;
+    publishInfo.payloadLength = strlen( MQTT_EXAMPLE_MESSAGE );
+
+    /* Get a new packet id. */
+    globalPublishPacketIdentifier = MQTT_GetPacketId( pContext );
+
+    /* Send PUBLISH packet. */
+    return MQTT_Publish( pContext,
+                         &publishInfo,
+                         globalPublishPacketIdentifier );
+}
+
+/* ============================   UNITY FIXTURES ============================ */
+
+/* Called before each test method. */
+void setUp()
+{
+    /* Reset file-scoped global variables. */
+    receivedSubAck = false;
+    receivedUnsubAck = false;
+    receivedPubAck = false;
+    receivedPubRec = false;
+    receivedPubRel = false;
+    receivedPubComp = false;
+    persistentSession = false;
+    memset( &incomingInfo, 0u, sizeof( MQTTPublishInfo_t ) );
+
+    /* Establish a TCP connection with the server endpoint. */
+    TEST_ASSERT_EQUAL( EXIT_SUCCESS,
+                       tcpConnectToServer( BROKER_ENDPOINT, BROKER_PORT, &tcpSocket ) );
+    TEST_ASSERT_NOT_EQUAL( -1, tcpSocket );
+
+    /* Establish TLS connection on top of TCP connection. */
+    TEST_ASSERT_EQUAL( EXIT_SUCCESS, tlsSetup( tcpSocket, SERVER_ROOT_CA_CERT_PATH, &pSslContext ) );
+    TEST_ASSERT_NOT_NULL( pSslContext );
+
+    /* Establish MQTT session on top of the TCP+TLS connection. */
+    establishMqttSession( &context, pSslContext, true, &persistentSession );
+}
+
+/* Called after each test method. */
+void tearDown()
+{
+    /* Free memory, if allocated during test case execution. */
+    if( incomingInfo.pTopicName != NULL )
+    {
+        free( ( void * ) incomingInfo.pTopicName );
+    }
+
+    if( incomingInfo.pPayload != NULL )
+    {
+        free( ( void * ) incomingInfo.pPayload );
+    }
+
+    /* Terminate MQTT connection. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, MQTT_Disconnect( &context ) );
```

Comment:
Are we supposed to have `TEST_ASSERT_EQUAL`s in a `tearDown()`? What happens if `MQTT_Disconnect` does not return success?

Reply:
Each test case represents a set of operations including connect, pub/sub, and disconnect, all of which are expected to pass for the test case to be considered successful. 
If disconnect fails, then we need to investigate the cause of the failure. 

I'm not sure I understand your concern with having assertions inside the `tearDown`